### DISCUSSION TITLE
Direct Disk Reads

### DISCUSSION THREAD

**Discussion by keshavvinayak01** [python-pillow/Pillow#6615] [Q&A] [answered]
Is there a way to use the Image.open API call with the O_DIRECT flag? 

I'm working on a project with application caching and I want to prevent all file-backed pages to load into the page cache (wasteful memory in my case) while calling the  Image.open API . Let me know where I could modify the code-base to add this specific feature, if possible.

Thanks.

**Comment by radarhere** [comment, marked as answer]
If you look at the source code for `Image.open`,
https://github.com/python-pillow/Pillow/blob/65c832d26e816e90c0cbb73bf3e09f91a92dd1cf/src/PIL/Image.py#L3050
the first parameter determines how the file is opened.
https://github.com/python-pillow/Pillow/blob/65c832d26e816e90c0cbb73bf3e09f91a92dd1cf/src/PIL/Image.py#L3060-L3063
If it is a filename, then Python is used to open the file, otherwise the file object passed in is used for subsequent read operations.
https://github.com/python-pillow/Pillow/blob/65c832d26e816e90c0cbb73bf3e09f91a92dd1cf/src/PIL/Image.py#L3099-L3109

See also https://pillow.readthedocs.io/en/stable/handbook/tutorial.html#reading-from-an-open-file

So I suspect you don't need to modify Pillow - you just need to open your file with `O_DIRECT` in Python, and then provide `Image.open` with an object to use to read data from it.

**Reply by keshavvinayak01** [reply]
This is what I tried as per your suggestion:

```python
def pil_loader(path):
    image_fd = os.open(path, os.O_RDONLY | os.O_DIRECT | os.O_SYNC)
    image_mmap = mmap.mmap(image_fd, os.path.getsize(path), offset=0, access=mmap.ACCESS_READ)
    img = Image.open(image_mmap)
    to_ret = img.convert('RGB')
    os.close(image_fd)
    return to_ret
````

This still leads to the size of buff/cache (Obtained from [free](https://man7.org/linux/man-pages/man1/free.1.html)) to increase, as well as the [vmtouch](https://hoytech.com/vmtouch/) utility also showed the images to be loaded into the cache.

Let me know what I should change.

**Reply by radarhere** [reply]
If you exclude Pillow for the moment, does the code you've provided perform in the way that you want?
```python
def pil_loader(path):
    image_fd = os.open(path, os.O_RDONLY | os.O_DIRECT | os.O_SYNC)
    image_mmap = mmap.mmap(image_fd, os.path.getsize(path), offset=0, access=mmap.ACCESS_READ)
    image_mmap.read()
```

If not, then this is a general Python problem, not something specific to Pillow.

**Reply by keshavvinayak01** [reply]
No it does not. Indeed it looks like a python specific problem, very strange.

**Reply by keshavvinayak01** [reply]
For future reference, I got it working. The following code achieves non-buffered reads:

```python3
image_fd = os.open(path, os.O_RDONLY | os.O_DIRECT)
fo = os.fdopen(image_fd, 'rb+', buffering=0)
image_mmap = mmap.mmap(-1, (1 << 24))
fo.readinto(image_mmap)
img = Image.open(image_mmap)
```

Can read about [fdopen](https://docs.python.org/3/library/os.html#os.fdopen) and [mmap](https://docs.python.org/3/library/mmap.html) to understand why this works, thanks :)

**Reply by radarhere** [reply]
Thanks for posting the solution!
